clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -598,16 +607,28 @@ private void failBatch(ProducerBatch batch, long baseOffset, long logAppendTime,
             } else if (transactionManager.isTransactional()) {
                 transactionManager.transitionToAbortableError(exception);
             }
+            transactionManager.stopTrackingBatch(batch);
+            if (adjustSequenceNumbers)
+                transactionManager.adjustSequencesDueToFailedBatch(batch);
```

Comment:
Discussed offline, but I guess the one case where it is not safe to adjust sequence numbers is when the number of retries are exhausted.